Config with default template a89af8f2

ale-rt opened the pull request.

ale-rt committed:
Configuring with plone/meta

ale-rt committed:
Run black

ale-rt committed:
Run pyupgrade

ale-rt committed:
Run isort

ale-rt committed:
Run zpretty

ale-rt committed:
Run codespell

ale-rt committed:
Run flake8

ale-rt committed:
Remove the obsolete travis.cfg

ale-rt committed:
Run check-manifest

ale-rt committed:
Require Python 3.8 and declare support for Plone >= 5.2

ale-rt committed:
Remove the obsolete getSiteEncoding call

ale-rt committed:
Configuring with plone/meta

ale-rt committed:
Update the code and the tests for Plone 6 and Python3

ale-rt committed:
Fix typo

ale-rt committed:
Do not run the dependecies check

ale-rt committed:
Comment the dependencies job

ale-rt committed:
Configuring with plone/meta removing the dependencies job

mauritsvanrees committed:
Add breaking news snippet.

mauritsvanrees committed:
Remove py_versions=3.8 from meta.toml.

Somehow it breaks GitHub Actions: the 'test' job is not run.
And `coverage` runs on 3.11 and passes, even though it shows an error in the tests.
Locally `tox -e test` fails with the same error.

mauritsvanrees committed:
Do not let the test browser handle errors.  Fix getting RichText output.

I want to see the actual error, instead of just getting a 500 Internal Server Error.
Now we get this error, which we fix in this same commit:

```
File "/Users/maurits/tmp/collective.watcherlist/collective/watcherlist/sample/newsletter.txt", line 107, in newsletter.txt
Failed example:
    browser.open(news.first.absolute_url() + '/@@newsitem-mail')
Exception raised:
    Traceback (most recent call last):
      File "/Users/maurits/.pyenv/versions/3.11.7/lib/python3.11/doctest.py", line 1353, in __run
        exec(compile(example.source, filename, "single",
      File "<doctest newsletter.txt[26]>", line 1, in <module>
        breakpoint(); browser.open(news.first.absolute_url() + '/@@newsitem-mail')
      File "/Users/maurits/tmp/collective.watcherlist/.tox/test/lib/python3.11/site-packages/zope/testbrowser/browser.py", line 252, in open
        self._processRequest(url, make_request)
      File "/Users/maurits/tmp/collective.watcherlist/.tox/test/lib/python3.11/site-packages/zope/testbrowser/browser.py", line 278, in _processRequest
        resp = make_request(reqargs)
      File "/Users/maurits/tmp/collective.watcherlist/.tox/test/lib/python3.11/site-packages/zope/testbrowser/browser.py", line 249, in make_request
        return self.testapp.get(url, **args)
      File "/Users/maurits/tmp/collective.watcherlist/.tox/test/lib/python3.11/site-packages/webtest/app.py", line 324, in get
        return self.do_request(req, status=status,
      File "/Users/maurits/tmp/collective.watcherlist/.tox/test/lib/python3.11/site-packages/zope/testbrowser/browser.py", line 92, in do_request
        response = super().do_request(req, status,
      File "/Users/maurits/tmp/collective.watcherlist/.tox/test/lib/python3.11/site-packages/webtest/app.py", line 620, in do_request
        res = req.get_response(app, catch_exc_info=True)
      File "/Users/maurits/tmp/collective.watcherlist/.tox/test/lib/python3.11/site-packages/webob/request.py", line 1309, in send
        status, headers, app_iter, exc_info = self.call_application(
      File "/Users/maurits/tmp/collective.watcherlist/.tox/test/lib/python3.11/site-packages/webob/request.py", line 1278, in call_application
        app_iter = application(self.environ, start_response)
      File "/Users/maurits/tmp/collective.watcherlist/.tox/test/lib/python3.11/site-packages/webtest/lint.py", line 196, in lint_app
        iterator = application(environ, start_response_wrapper)
      File "/Users/maurits/tmp/collective.watcherlist/.tox/test/lib/python3.11/site-packages/plone/testing/_z2_testbrowser.py", line 36, in wrapped_func
        return func(*args, **kw)
      File "/Users/maurits/tmp/collective.watcherlist/.tox/test/lib/python3.11/site-packages/plone/testing/_z2_testbrowser.py", line 63, in __call__
        wsgi_result = publish(environ, start_response)
      File "/Users/maurits/tmp/collective.watcherlist/.tox/test/lib/python3.11/site-packages/ZPublisher/WSGIPublisher.py", line 391, in publish_module
        response = _publish(request, new_mod_info)
      File "/Users/maurits/tmp/collective.watcherlist/.tox/test/lib/python3.11/site-packages/ZPublisher/WSGIPublisher.py", line 285, in publish
        result = mapply(obj,
      File "/Users/maurits/tmp/collective.watcherlist/.tox/test/lib/python3.11/site-packages/ZPublisher/mapply.py", line 98, in mapply
        return debug(object, args, context)
      File "/Users/maurits/tmp/collective.watcherlist/.tox/test/lib/python3.11/site-packages/ZPublisher/WSGIPublisher.py", line 68, in call_object
        return obj(*args)
      File "/Users/maurits/tmp/collective.watcherlist/collective/watcherlist/browser.py", line 54, in __call__
        return self.html
      File "/Users/maurits/tmp/collective.watcherlist/collective/watcherlist/sample/browser.py", line 18, in html
        return text.output()
    TypeError: 'str' object is not callable
```

mauritsvanrees commented on collective/watcherlist/sample/newsletter.txt:
```diff
@@ -182,23 +191,20 @@ Note the different charset and encoding::
 for debugging purposes)::
 
-    >>> ori_smtp_host = self.get_smtp_host()
-    >>> self.configure_mail_host(u'')
```
I think we do still want these lines, otherwise the test is invalid.
But this file has test errors anyway, so needs another look.

mauritsvanrees replied:
Fixed.

mauritsvanrees reviewed (commented):
Thanks!  Mostly fine.

But locally `tox -e test` fails.  And in the actions you don't notice this, because the `test` job is never executed.  I suspect there is something not working when you select `py_versions` in meta toml.

I added a few small commits, fixing this and the first error.
I will have a further look, because it seems either the mock mail host setup in the tests is not right, or the way we get the mailhost in the utils is not right.

mauritsvanrees committed:
manifest: do not include the news directory.

mauritsvanrees commented:
This is weird.  GitHub Actions has [an error](https://github.com/collective/collective.watcherlist/actions/runs/9676689067/job/26696814983?pr=14#step:8:159) but it still is marked as green.

Also locally with `tox -e test` I see in the output that there is a failure, but this is not counted in the test summary, and the exit code is zero.

I suspect this might be because it is in `newsletter.txt` in the `sample` module.  I will see if I can move this file to "fix" it.

mauritsvanrees committed:
Fix test_integration: test failures were not reported.

I took some hints from `plone/app/dexterity/tests/test_doctests.py`.

Note that you do get some test failures now.

mauritsvanrees commented:
I have fixed `test_integration.py`.  Now the test failures are at least reported correctly.

mauritsvanrees committed:
Tests: do not publish the two news items.

We want to see what happens when we publish them later, as then we want to send mails.

mauritsvanrees commented:
In `test_integration.py` we should not have published the two news items.  In `newsletter.txt` we publish them to let a sample event handler send mails.  Fixed now in latest commit.
Still it fails: the mock mail host has no messages, even though when I go in with a pdb I do see mails getting added to the messages list when we publish the first item.  A transaction commit does not help.  Not sure yet here.
Now picking up my kids. :-)

mauritsvanrees committed:
Fix functional tests by using a browser view to check the mailhost messages.

mauritsvanrees reviewed (approved):
I have fixed the tests.
@ale-rt Can you have a look at my changes?

Note that this is not tested on Plone 5.2 automatically, that is not yet easily possibly with plone/meta.  But I suppose it should work.

ale-rt commented on collective/watcherlist/tests/test_integration.py:
```diff
@@ -132,9 +132,7 @@ def browser_login(self, user=None):
 
 
-OPTIONFLAGS = (
-    doctest.ELLIPSIS | doctest.NORMALIZE_WHITESPACE | doctest.REPORT_ONLY_FIRST_FAILURE
```
Yeah, I forgot to remove it after debugging the tests :p

ale-rt reviewed (commented):
LGTM

ale-rt commented:
@mauritsvanrees could you please release an alpha version?
Then we can start live testing this and extend the Plone/Python matrix.

mauritsvanrees commented:
@ale-rt I have released 4.0.0a1.
